src/attacks.rs

```diff
@@ -1,27 +1,209 @@
+pub struct GreedyParams {
+    // how many k nodes do we "remove" at each iteration in append_removal
+    pub k: usize,
+    // the radius for the heuristic to delete as well close nodes within a
+    // radius of a selected node.
+    pub radius: usize,
+    //#[allow(dead_code)]
+    // maximum lenth of the path - heuristic for the table produced by count_paths
+    // see paragraph below equation 8.
+    // TODO: NOT (really) IMPLEMENTED YET - careful work is required
+    // pub length: usize,
+}
+// greedy_reduce implements the Algorithm 5 of https://eprint.iacr.org/2018/944.pdf
+fn greedy_reduce(g: &Graph, target: usize, p: GreedyParams) -> HashSet<usize> {
+    let mut s = HashSet::new();
+    let mut inradius: HashSet<usize> = HashSet::new();
+    while g.depth_exclude(&s) > target {
+        // TODO use p.length when more confidence in the trick
+        let (_, topk) = count_paths(g, &s, target, p.k);
+        append_removal(g, &mut s, &topk, &mut inradius, p.radius);
+    }
+    s
+}
+
+// append_removal is an adaptation of "SelectRemovalNodes" function in Algorithm 6
+// of https://eprint.iacr.org/2018/944.pdf. Instead of returning the set of nodes
+// to remove, it simply adds them to the given set.
+fn append_removal(
+    g: &Graph,
+    set: &mut HashSet<usize>,
+    topk: &Vec<Pair>,
+    inradius: &mut HashSet<usize>,
+    radius: usize,
+) {
+    if radius == 0 {
+        // take the node with the highest number of incident path
+        set.insert(topk.iter().max_by_key(|pair| pair.1).unwrap().0);
+        return;
+    }
+
+    let mut unseen = topk
+        .iter()
+        // take the nodes that are not yet in the inradius set
+        .filter(|&pair| !inradius.contains(&pair.0))
+        .collect::<Vec<&Pair>>();
+
+    // if all nodes are already in the radius set, then at least take
+    // the first one.
+    // https://github.com/filecoin-project/drg-attacks/issues/2
+    // for more details.
+    if unseen.len() == 0 {
+        unseen.push(&topk[0]);
+    }
+
+    for &pair in unseen.iter() {
+        set.insert(pair.0);
+        update_radius_set(g, pair.0, inradius, radius);
+    }
+}
+
+// update_radius_set fills the given inradius set with nodes that inside a radius
+// of the given node. Size of the radius is given radius. It corresponds to the
+// under-specified function "UpdateNodesInRadius" in algo. 6 of
+// https://eprint.iacr.org/2018/944.pdf
+fn update_radius_set(g: &Graph, node: usize, inradius: &mut HashSet<usize>, radius: usize) {
+    let add_direct_nodes = |v: usize, closests: &mut Vec<usize>| {
+        // add all direct parent
+        g.parents()[v]
+            .iter()
+            .for_each(|&parent| closests.push(parent));
 
-// valiant_basic returns a set S such that depth(G - S) < target.
+        // add all direct children
+        // TODO: compute once children graph and use it again as in C# instead
+        // of searching linearly
```

Comment:
Yes, this sounds like it might make a substantial difference.

Also, in terms of improving the efficiency a bit, the radius search should be split between parents and children directions, once I go looking in one direction I can't go in the reverse (since we look for directed paths), e.g., if I'm tracing node 5, and then I add its parent, node 3, the next iteration will look for the children of 3, say, 7, but that doesn't mean there's a path of length 2 between nodes 5 and 7 (just that they have a common ancestor, node 3), in that search path I should continue asking only for the parents (and similarly when looking for the children). Does that make sense?